### PR TITLE
Validate plugin source set defined in extension

### LINKED ISSUE #10759: Plugin validation should validate plugin source set
Currently `JavaGradlePluginPlugin` uses the `GradlePluginDevelopmentExtension` extension's `pluginSourceSet` to determine where the plugin sources can be found. However, the `validateTaskProperties` task registered automatically for the plugin sources uses the main sourceset.

This probably doesn't cause any problems because the `pluginSourceSet` defaults to the main source set, and hardly anyone bothers to change that.

Still, the auto-registered task should validate `pluginSourceSet` instead.

We should probably use a task naming scheme similar to how we name compile tasks to avoid collisions. I.e. `validateTaskProperties` should be validating the `main` source set, while `validateXxxTaskProperties` should validate things defined in the `xxx` source set.

The latter scheme would work even better after we've renamed the validator task to `ValidatePlugins` and the corresponding tasks to `validatePlugins` (see #10546 and #10757).

---
cc: @gradle/build-cache

### PR DESCRIPTION
**Pull request by lptr**
Fixes #10759

### CODE REVIEW

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java [hunk 1]
```diff
@@ -238,8 +238,8 @@ private void validatePluginDeclarations(Project project, GradlePluginDevelopment
         });
     }
 
-    private void configurePluginValidations(Project project) {
-        SourceSet mainSourceSet = project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+    private void configurePluginValidations(Project project, GradlePluginDevelopmentExtension extension) {
+        SourceSet pluginSourceSet = extension.getPluginSourceSet();
```
Comment (wolfs) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java]: Shouldn't this lookup be lazy? If we do it here, then changing the `pluginSourceSet` on the extension won't have any effect, right?

Let's also add a test for this new behaviour.

Comment (lptr) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java]: Well spotted. Added test and fix.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java [hunk 2]
```diff
@@ -256,9 +255,9 @@ public void execute(Task taskWithAction) {
 
             task.getOutputFile().set(project.getLayout().getBuildDirectory().file("reports/plugin-development/validation-report.txt"));
 
-            task.getClasses().setFrom(mainSourceSet.getOutput().getClassesDirs());
-            task.getClasspath().setFrom(mainSourceSet.getCompileClasspath());
-            task.dependsOn(mainSourceSet.getOutput());
+            task.getClasses().setFrom((Callable<Object>) () -> extension.getPluginSourceSet().getOutput().getClassesDirs());
+            task.getClasspath().setFrom((Callable<Object>) () -> extension.getPluginSourceSet().getCompileClasspath());
+            task.dependsOn((Callable<Object>) () -> extension.getPluginSourceSet().getOutput());
```
Comment (wolfs) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java]: Isn't the dependency automatically added by the classes? Given that `classesDirs.builtBy(this)`? https://github.com/gradle/gradle/blob/8a082ed93daba60c41b4687b429ac9a3c0397464/subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSetOutput.java#L49-L51

Comment (lptr) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java]: Good point, I removed it.

File: subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/JavaGradlePluginPluginTestKitSetupIntegrationTest.groovy [hunk 3]
```diff
@@ -73,6 +73,7 @@ class JavaGradlePluginPluginTestKitSetupIntegrationTest extends AbstractIntegrat
                 custom {
                     java {
                         srcDir 'src'
+                        compileClasspath = configurations.compileClasspath
```
Comment (wolfs) [inline subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/plugins/JavaGradlePluginPluginTestKitSetupIntegrationTest.groovy]: How did this work before?

**Pull request merged**